### PR TITLE
[SPARK-37585][SQL] Update InputMetric in DataSourceRDD with TaskCompletionListener

### PR DESCRIPTION
**Pull request by bozhang2820**
### What changes were proposed in this pull request?
Before this change, Spark only updates `InputMetrics.bytesRead` in `DataSourceRDD` once every 1000 records, or at the end of the iterator (when `MetricsIterator.hasNext` returns false). So when the output is limited in a query but there is still data in the datasource, `InputMetrics.bytesRead` will not be updated at the end of the iterator, leading to incorrect metric results. This is more pronounced when the total number of records is less than 1000 (which will lead to `InputMetrics.bytesRead == 0`).

This change fixes this bug by adding a force metric update in `TaskCompletionListener`.

### Why are the changes needed?
This is to fix the bug that `InputMetrics.bytesRead` is not updated when there is still data in the datasource but the output is limited. 

### Does this PR introduce _any_ user-facing change?
Users will see more accurate `InputMetrics.bytesRead` with this change.

### How was this patch tested?
Added a unit test.

### CODE REVIEW

**Comment by bozhang2820** [comment]
CC @cloud-fan for a review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala [hunk 1]
```diff
@@ -126,14 +127,9 @@ private class MetricsHandler extends Logging with Serializable {
...
 
-  override def hasNext: Boolean = {
-    if (iter.hasNext) {
-      true
-    } else {
-      metricsHandler.updateMetrics(0, force = true)
-      false
-    }
-  }
+  override def hasNext: Boolean = iter.hasNext
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala:130]: Can we keep updating metrics at the end of the iterator? It's still useful in cases like `RDD.coalesce`, where a task consumes multiple iterators for coalesced partitions, and we can update metrics earlier here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 2]
```diff
@@ -244,4 +244,33 @@ class DataSourceV2ScanExecRedactionSuite extends DataSourceScanRedactionTest {
       }
     }
   }
+
+  test("SPARK-37585: test input metrics for DSV2 with output limits") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: I should have pointed it out earlier... It's weird to put metrics test in this suite, how about `FileBasedDataSourceSuite`?

**Comment by cloud-fan** [comment]
also cc @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala (hunk 1 above)
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala:130]: Sure.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala (hunk 2 above)
Comment (bozhang2820) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: Sure, but should I move the other related tests together in this PR as well?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: don't you have only one new test? any other related tests?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala [hunk 3]
```diff
@@ -70,6 +70,7 @@ class DataSourceRDD(
       // In case of early stopping before consuming the entire iterator,
       // we need to do one more metric update at the end of the task.
       CustomMetrics.updateMetrics(reader.currentMetricsValues, customMetrics)
+      iter.forceUpdateMetrics()
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala:73]: looks okay.

**Review by viirya — COMMENTED** [review]
Please don't forget to update the PR description as this still keeps updating metrics at the end of the iterator.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala (hunk 2 above)
Comment (bozhang2820) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: I meant the other test related to InputMetrics here.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: yea let's move them together

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:248]: oh i see. yea, okay to move that together.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**